general_messaging_constructs.md

```diff
@@ -4,9 +4,75 @@
+
+We need a way to "attach" such content to DIDComm messages. This method must be flexible, powerful, and usable without requiring new schema updates for every dynamic variation.
+
+### Reference
+
+Attachments are contained within a list in the `attachments` header.
+
+Each attachment is contained within the following structure.
+
+- `id`: Identifies attached content within the scope of a given message. Recommended on appended attachment descriptors. Possible but generally unused on embedded attachment descriptors. Never required if no references to the attachment exist; if omitted, then there is no way to refer to the attachment later in the thread, in error messages, and so forth. Because `id` is used to compose URIs, it is recommended that this name be brief and avoid spaces and other characters that require URI escaping.
+- `description`: An optional human-readable description of the content.
+- `filename`: A hint about the name that might be used if this attachment is persisted as a file. It is not required, and need not be unique. If this field is present and `mime-type` is not, the extension on the filename may be used to infer a MIME type.
+- `mime_type`: Describes the MIME type of the attached content. Optional but recommended.
+- `lastmod_time`: A hint about when the content in this attachment was last modified.
+- `byte_count`: Optional, and mostly relevant when content is included by reference instead of by value. Lets the receiver guess how expensive it will be, in time, bandwidth, and storage, to fully fetch the attachment.
+- `data`: A JSON object that gives access to the actual content of the attachment. Contains the following subfields:
+  - `jws`: A JSON Web Signature over the content of the attachment. Optional.
+  - `hash`: The hash of the content encoded in multi-hash format. Optional. Used as an integrity check for the attachment, and MUST be used if the data is referenced via the `links` data attribute. 
+  - `links`: A list of zero or more locations at which the content may be fetched. Optional.
+  - `base64`: Base64-encoded data, when representing arbitrary content inline instead of via `links`. Optional.
+  - `encrypted`: Already encrypted data, encoded in Base64. Data contained in this field MUST be encrypted.
+  - `json`: Directly embedded JSON data, when representing content inline instead of via `links`, and when the content is natively conveyable as JSON. Optional.
+
+### Example
+
+```json
+{
+    "type": "<sometype>",
+    "to": ["did:example:mediator"],
+    "body":{
+        "attachment_id": "1",
+        "encrypted_details": {
+            "id": "x",
+            "encrypted_to": "",
+            "other_details": "about attachment"
+        }
+    },
+    "attachments": [
```

Comment:
What's the main advantage of keeping the attachment outside the body of the payload? @dhh1128 might be able to point out a few, but I'm not remembering from when this was originally designed.

Reply:
I think this facilitates a solution to avoid re-encrypting data that's already been encrypted once. @TelegramSam ?

Reply:
Wouldn't the attachment data still be placed inside the encryption envelop though since it's a part of the JWM?

Reply:
More on this in today's meeting.